File: app/src/main/java/com/raival/fileexplorer/tab/file/FileExplorerTabFragment.java
package com.raival.fileexplorer.tab.file;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Environment;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textfield.TextInputLayout;
import com.raival.fileexplorer.App;
import com.raival.fileexplorer.R;
import com.raival.fileexplorer.activity.MainActivity;
import com.raival.fileexplorer.activity.TextEditorActivity;
import com.raival.fileexplorer.common.dialog.CustomDialog;
import com.raival.fileexplorer.tab.BaseDataHolder;
import com.raival.fileexplorer.tab.BaseTabFragment;
import com.raival.fileexplorer.tab.checklist.ChecklistTabFragment;
import com.raival.fileexplorer.tab.file.adapter.FileListAdapter;
import com.raival.fileexplorer.tab.file.adapter.PathRootAdapter;
import com.raival.fileexplorer.tab.file.dialog.SearchDialog;
import com.raival.fileexplorer.tab.file.dialog.TasksDialog;
import com.raival.fileexplorer.tab.file.executor.DexRunner;
import com.raival.fileexplorer.tab.file.model.FileItem;
import com.raival.fileexplorer.tab.file.option.FileOptionHandler;
import com.raival.fileexplorer.tab.file.util.FileExtensions;
import com.raival.fileexplorer.tab.file.util.FileUtils;
import com.raival.fileexplorer.util.PrefsUtils;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

public class FileExplorerTabFragment extends BaseTabFragment {
    public final static int MAX_NAME_LENGTH = 32;
    private final ArrayList<FileItem> files = new ArrayList<>();
    private RecyclerView fileList;
    private RecyclerView pathRootRv;
    private View placeHolder;
    private FileOptionHandler fileOptionHandler;
    private File previousDirectory;
    private File currentDirectory;

    private boolean requireRefresh = false;

    private IconResolver iconResolver;
    private Thread iconResolverThread;

    public FileExplorerTabFragment() {
        super();
    }

    public FileExplorerTabFragment(File directory) {
        super();
        currentDirectory = directory;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.file_explorer_tab_fragment, container, false);
        fileList = view.findViewById(R.id.rv);
        pathRootRv = view.findViewById(R.id.path_root);
        placeHolder = view.findViewById(R.id.place_holder);

        view.findViewById(R.id.home).setOnClickListener(view1 -> setCurrentDirectory(getDefaultHomeDirectory()));
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        prepareBottomBarView();
        initFileList();
        loadData();
        // restore RecyclerView state
        restoreRecyclerViewState();
    }

    private void loadData() {
        setCurrentDirectory(((FileExplorerTabDataHolder) getDataHolder()).activeDirectory);
    }

    public void prepareBottomBarView() {
        getBottomBarView().clear();
        getBottomBarView().addItem("Tasks", R.drawable.ic_baseline_assignment_24, (view) ->
                new TasksDialog(this).show(getParentFragmentManager(), ""));

        getBottomBarView().addItem("Search", R.drawable.ic_round_search_24, view -> {
            SearchDialog searchFragment = new SearchDialog(this, getCurrentDirectory());
            searchFragment.show(getParentFragmentManager(), "");
            setSelectAll(false);
        });
        getBottomBarView().addItem("Create", R.drawable.ic_baseline_add_24, view -> showAddNewFileDialog());
        getBottomBarView().addItem("Sort", R.drawable.ic_baseline_sort_24, this::showSortOptionsMenu);
        getBottomBarView().addItem("Select All", R.drawable.ic_baseline_select_all_24, view -> setSelectAll(true));
        getBottomBarView().addItem("refresh", R.drawable.ic_baseline_restart_alt_24, view -> refresh());
    }

    private void showSortOptionsMenu(View view) {
        PopupMenu popupMenu = new PopupMenu(requireActivity(), view);

        popupMenu.getMenu().add("Sort by:").setEnabled(false);

        popupMenu.getMenu().add("Name (A-Z)").setCheckable(true).setChecked(PrefsUtils.getSortingMethod() == PrefsUtils.SORT_NAME_A2Z);
        popupMenu.getMenu().add("Name (Z-A)").setCheckable(true).setChecked(PrefsUtils.getSortingMethod() == PrefsUtils.SORT_NAME_Z2A);

        popupMenu.getMenu().add("Size (Bigger)").setCheckable(true).setChecked(PrefsUtils.getSortingMethod() == PrefsUtils.SORT_SIZE_BIGGER);
        popupMenu.getMenu().add("Size (Smaller)").setCheckable(true).setChecked(PrefsUtils.getSortingMethod() == PrefsUtils.SORT_SIZE_SMALLER);

        popupMenu.getMenu().add("Date (Newer)").setCheckable(true).setChecked(PrefsUtils.getSortingMethod() == PrefsUtils.SORT_DATE_NEWER);
        popupMenu.getMenu().add("Date (Older)").setCheckable(true).setChecked(PrefsUtils.getSortingMethod() == PrefsUtils.SORT_DATE_OLDER);

        popupMenu.getMenu().add("Other options:").setEnabled(false);

        popupMenu.getMenu().add("Folders first").setCheckable(true).setChecked(PrefsUtils.listFoldersFirst());

        popupMenu.setOnMenuItemClickListener(menuItem -> {
            menuItem.setChecked(!menuItem.isChecked());
            switch (menuItem.getTitle().toString()) {
                case "Name (A-Z)": {
                    PrefsUtils.setSortingMethod(PrefsUtils.SORT_NAME_A2Z);
                    break;
                }
                case "Name (Z-A)": {
                    PrefsUtils.setSortingMethod(PrefsUtils.SORT_NAME_Z2A);
                    break;
                }
                case "Size (Bigger)": {
                    PrefsUtils.setSortingMethod(PrefsUtils.SORT_SIZE_BIGGER);
                    break;
                }
                case "Size (Smaller)": {
                    PrefsUtils.setSortingMethod(PrefsUtils.SORT_SIZE_SMALLER);
                    break;
                }
                case "Date (Older)": {
                    PrefsUtils.setSortingMethod(PrefsUtils.SORT_DATE_OLDER);
                    break;
                }
                case "Date (Newer)": {
                    PrefsUtils.setSortingMethod(PrefsUtils.SORT_DATE_NEWER);
                    break;
                }
                case "Folders first": {
                    PrefsUtils.setListFoldersFirst(menuItem.isChecked());
                    break;
                }
            }
            refresh();
            return true;
        });
        popupMenu.show();
    }

    private void showAddNewFileDialog() {
        @SuppressLint("InflateParams") TextInputLayout input = (TextInputLayout) getLayoutInflater().inflate(R.layout.input, null, false);
        input.setHint("File name");
        Objects.requireNonNull(input.getEditText()).setSingleLine();
        FileUtils.setFileValidator(input, getCurrentDirectory());

        new CustomDialog()
                .setTitle("Create new file")
                .addView(input)
                .setPositiveButton("File", view ->
                        createFile(input.getEditText().getText().toString(), false), true)
                .setNegativeButton("Folder", view ->
                        createFile(input.getEditText().getText().toString(), true), true)
                .setNeutralButton("Cancel", null, true)
                .show(getParentFragmentManager(), "");
    }

    public void createFile(String name, boolean isFolder) {
        File file = new File(getCurrentDirectory(), name);
        if (isFolder) {
            if (!file.mkdir()) {
                App.showMsg("Unable to create folder: " + file.getAbsolutePath());
            } else {
                refresh();
                focusOn(file);
            }
        } else {
            try {
                if (!file.createNewFile()) {
                    App.showMsg("Unable to create file: " + file.getAbsolutePath());
                } else {
                    refresh();
                    focusOn(file);
                }
            } catch (IOException e) {
                App.showMsg(e.toString());
                App.log(e);
            }
        }
    }

    @Override
    public void closeTab() {
        // Close the tab (if not default tab)
        if (getTag() != null && !getTag().startsWith("0_")) {
            super.closeTab();
        }
    }

    @Override
    public boolean onBackPressed() {
        //  Unselect selected files (if any)
        if (getSelectedFiles().size() > 0) {
            setSelectAll(false);
            return true;
        }
        // Go back if possible
        final File parent = getCurrentDirectory().getParentFile();
        if (parent != null && parent.exists() && parent.canRead()) {
            setCurrentDirectory(getCurrentDirectory().getParentFile());
            // restore RecyclerView state
            restoreRecyclerViewState();
            // Clear any selected files from the DataHolder (it also gets cleared
            // when a FileItem is clicked)
            ((FileExplorerTabDataHolder) getDataHolder()).selectedFiles.clear();
            return true;
        }
        // Close the tab (if not default tab)
        if (getTag() != null && !getTag().startsWith("0_")) {
            closeTab();
            return true;
        }

        return false;
    }

    @Override
    public BaseDataHolder createNewDataHolder() {
        FileExplorerTabDataHolder _dataHolder = new FileExplorerTabDataHolder(getTag());
        _dataHolder.activeDirectory = currentDirectory == null ? getDefaultHomeDirectory() : currentDirectory;
        return _dataHolder;
    }

    @Override
    public void onStop() {
        super.onStop();
        requireRefresh = true;
    }

    @Override
    public void onPause() {
        super.onPause();
        requireRefresh = true;
    }

    @Override
    public void onResume() {
        super.onResume();
        if (requireRefresh) {
            requireRefresh = false;
            refresh();
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setSelectAll(boolean select) {
        if (!select) ((FileExplorerTabDataHolder) getDataHolder()).selectedFiles.clear();

        for (FileItem item : files) {
            item.isSelected = select;
            if (select) {
                ((FileExplorerTabDataHolder) getDataHolder()).selectedFiles.add(item.file);
            }
        }
        // Don't call refresh(), because it will recreate the tab and reset the selection
        Objects.requireNonNull(fileList.getAdapter()).notifyDataSetChanged();
    }

    public ArrayList<FileItem> getSelectedFiles() {
        final ArrayList<FileItem> list = new ArrayList<>();
        for (FileItem item : files) {
            if (item.isSelected) list.add(item);
        }
        return list;
    }

    /**
     * Show/Hide placeholder
     */
    public void showPlaceholder(boolean isShow) {
        placeHolder.setVisibility(isShow ? View.VISIBLE : View.GONE);
    }

    /**
     * Used to update the title of attached tabView
     */
    private void updateTabTitle() {
        getTabView().setName(getName());
    }

    /**
     * This method is called once from #onViewCreated(View, Bundle)
     */
    private void initFileList() {
        fileList.setAdapter(new FileListAdapter(this));
        fileList.setHasFixedSize(true);
        initPathRoot();
    }

    private void initPathRoot() {
        pathRootRv.setLayoutManager(new LinearLayoutManager(requireActivity(), LinearLayoutManager.HORIZONTAL, false));
        pathRootRv.setAdapter(new PathRootAdapter(this));
    }

    /**
     * RecyclerView state should be saved when the fragment is destroyed and recreated.
     * #getDataHolder() isn't used here because we don't want to create a new DataHolder if the fragment is about
     * to close (note that the DataHolder gets removed just right before the fragment is closed)
     */
    @Override
    public void onDestroy() {
        super.onDestroy();
        FileExplorerTabDataHolder fileExplorerTabDataHolder = (FileExplorerTabDataHolder) getMainViewModel().getDataHolder(getTag());
        if (fileExplorerTabDataHolder != null) {
            fileExplorerTabDataHolder.recyclerViewStates.put(getCurrentDirectory(), Objects.requireNonNull(fileList.getLayoutManager()).onSaveInstanceState());
        }
    }

    /**
     * This method automatically removes the restored state from DataHolder recyclerViewStates
     * This method is called when:
     * - Create the fragment
     * - #onBackPressed()
     * - when select a directory from pathRoot RecyclerView
     */
    public void restoreRecyclerViewState() {
        Parcelable savedState = ((FileExplorerTabDataHolder) getDataHolder()).recyclerViewStates.get(getCurrentDirectory());
        if (savedState != null) {
            Objects.requireNonNull(fileList.getLayoutManager()).onRestoreInstanceState(savedState);
            ((FileExplorerTabDataHolder) getDataHolder()).recyclerViewStates.remove(getCurrentDirectory());
        }
    }

    /**
     * Refreshes both fileList and pathRoot recyclerview (used by #setCurrentDirectory(File) ONLY)
     */
    @SuppressLint("NotifyDataSetChanged")
    private void refreshFileList() {
        Objects.requireNonNull(fileList.getAdapter()).notifyDataSetChanged();
        Objects.requireNonNull(pathRootRv.getAdapter()).notifyDataSetChanged();
        pathRootRv.scrollToPosition(pathRootRv.getAdapter().getItemCount() - 1);
        fileList.scrollToPosition(0);
        if (getToolbar() != null)
            getToolbar().setSubtitle(FileUtils.getFormattedFileCount(getCurrentDirectory()));
    }

    /**
     * Used to refresh the tab
     */
    public void refresh() {
        setCurrentDirectory(getCurrentDirectory());
        restoreRecyclerViewState();
    }

    private File getDefaultHomeDirectory() {
        return Environment.getExternalStorageDirectory();
    }

    private void prepareFiles() {
        // Make sure current file is ready
        if (getCurrentDirectory() == null) {
            loadData();
            return;
        }
        // Clear previous list
        files.clear();
        // Load all files in the current File
        File[] files = getCurrentDirectory().listFiles();
        if (files != null) {
            for (Comparator<File> comparator : FileUtils.getComparators()) {
                Arrays.sort(files, comparator);
            }
            for (File file : files) {
                FileItem fileItem = new FileItem(file);
                if (((FileExplorerTabDataHolder) getDataHolder()).selectedFiles.contains(fileItem.file)) {
                    fileItem.isSelected = true;
                }
                this.files.add(fileItem);
            }
        }
        iconResolver = new IconResolver().start();
    }

    public void focusOn(File file) {
        for (int i = 0; i < files.size(); i++) {
            if (file.equals(files.get(i).file)) {
                fileList.scrollToPosition(i);
                return;
            }
        }
    }

    /**
     * @return the name associated with this tab (currently used for tabView)
     */
    public String getName() {
        return FileUtils.getShortLabel(getCurrentDirectory(), MAX_NAME_LENGTH);
    }

    public void showFileOptions(FileItem fileItem) {
        if (fileOptionHandler == null) {
            fileOptionHandler = new FileOptionHandler(this);
        }
        fileOptionHandler.showOptions(fileItem);
    }

    public void openFile(FileItem fileItem) {
        if (!handleKnownFileExtensions(fileItem)) {
            FileUtils.openFileWith(fileItem.file, false);
        }
    }

    private boolean handleKnownFileExtensions(FileItem fileItem) {
        if (FileUtils.isTextFile(fileItem.file) || FileUtils.isCodeFile(fileItem.file)) {
            Intent intent = new Intent();
            intent.setClass(requireActivity(), TextEditorActivity.class);
            intent.putExtra("file", fileItem.file.getAbsolutePath());
            requireActivity().startActivity(intent);
            return true;
        }
        if (FileUtils.getFileExtension(fileItem.file).equals("checklist")) {
            ((MainActivity) requireActivity()).addNewTab(new ChecklistTabFragment(fileItem.file)
                    , "ChecklistTabFragment_" + ((MainActivity) requireActivity()).generateRandomTag());
            return true;
        }
        if (fileItem.file.getName().toLowerCase().endsWith(".exe.dex")) {
            new DexRunner(fileItem.file, (AppCompatActivity) requireActivity()).run();
            return true;
        }
        return false;
    }

    public void showDialog(String title, String msg) {
        new CustomDialog()
                .setTitle(title)
                .setMsg(msg)
                .setPositiveButton("Ok", null, true)
                .showDialog(getParentFragmentManager(), "");
    }

    public ArrayList<FileItem> getFiles() {
        return files;
    }

    //______________| Getter and Setter |_______________\\

    public File getCurrentDirectory() {
        return currentDirectory;
    }

    /**
     * This method handles the following (in order):
     * - Updating currentDirectory and previousDirectory fields
     * - Updating recyclerViewStates in DataHolder
     * - Sorting files based on the preferences
     * - Updating tabView title
     * - Refreshing adapters (fileList & pathRoot)
     * - Updating activeDirectory in DataHolder
     *
     * @param dir the directory to open
     */
    public void setCurrentDirectory(File dir) {
        previousDirectory = currentDirectory;
        currentDirectory = dir;
        // Save only when previousDirectory is set (so that it can restore the state before onDestroy())
        if (previousDirectory != null) ((FileExplorerTabDataHolder) getDataHolder())
                .recyclerViewStates.put(previousDirectory, Objects.requireNonNull(fileList.getLayoutManager()).onSaveInstanceState());
        prepareFiles();
        updateTabTitle();
        refreshFileList();
        ((FileExplorerTabDataHolder) getDataHolder()).activeDirectory = getCurrentDirectory();
    }

    public File getPreviousDirectory() {
        return previousDirectory;
    }

    private class IconResolver {
        public boolean isWorking = false;

        public IconResolver start() {
            isWorking = true;
            AtomicInteger i = new AtomicInteger();
            iconResolverThread = new Thread(() -> {
                for (FileItem fileItem : files) {
                    if (FileUtils.getFileExtension(fileItem.file).equalsIgnoreCase(FileExtensions.apkType)) {
                        Drawable drawable = FileUtils.getApkIcon(fileItem.file);
                        App.appHandler.post(() -> fileItem.img.setValue(drawable));
                    }
                    i.incrementAndGet();
                }
                isWorking = false;
            });
            iconResolverThread.start();
            return this;
        }
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/FileExplorerTabFragment.kt
package com.raival.fileexplorer.tab.file

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textview.MaterialTextView
import com.raival.fileexplorer.App.Companion.showMsg
import com.raival.fileexplorer.R
import com.raival.fileexplorer.activity.MainActivity
import com.raival.fileexplorer.common.dialog.CustomDialog
import com.raival.fileexplorer.extension.getFormattedFileCount
import com.raival.fileexplorer.extension.getShortLabel
import com.raival.fileexplorer.extension.toDp
import com.raival.fileexplorer.tab.BaseDataHolder
import com.raival.fileexplorer.tab.BaseTabFragment
import com.raival.fileexplorer.tab.file.adapter.FileListAdapter
import com.raival.fileexplorer.tab.file.adapter.PathHistoryAdapter
import com.raival.fileexplorer.tab.file.dialog.SearchDialog
import com.raival.fileexplorer.tab.file.dialog.TasksDialog
import com.raival.fileexplorer.tab.file.misc.FileOpener
import com.raival.fileexplorer.tab.file.misc.FileUtils
import com.raival.fileexplorer.tab.file.model.FileItem
import com.raival.fileexplorer.tab.file.options.FileOptionsHandler
import com.raival.fileexplorer.util.Log
import com.raival.fileexplorer.util.PrefsUtils
import java.io.File
import java.io.IOException
import java.util.*

class FileExplorerTabFragment : BaseTabFragment {
    val files = ArrayList<FileItem>()
    var pathHistory = ArrayList<File>()

    private lateinit var fileList: RecyclerView
    private lateinit var pathHistoryRv: RecyclerView
    private lateinit var placeHolder: View
    private lateinit var fileOptionsHandler: FileOptionsHandler
    private var requireRefresh = false
    var previousDirectory: File? = null

    @JvmField
    var currentDirectory: File? = null

    constructor() : super()
    constructor(directory: File) : super() {
        currentDirectory = directory
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.file_explorer_tab_fragment, container, false)
        fileList = view.findViewById(R.id.rv)
        pathHistoryRv = view.findViewById(R.id.path_history)
        placeHolder = view.findViewById(R.id.place_holder)
        val homeButton = view.findViewById<View>(R.id.home)
        homeButton.setOnClickListener {
            setCurrentDirectory(defaultHomeDirectory)
        }
        homeButton.setOnLongClickListener { showSetPathDialog() }
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        prepareBottomBarView()
        initFileList()
        loadData()
        // restore RecyclerView state
        restoreRecyclerViewState()
    }

    private fun loadData() {
        setCurrentDirectory((dataHolder as FileExplorerTabDataHolder).activeDirectory!!)
    }

    private fun prepareBottomBarView() {
        bottomBarView!!.clear()
        bottomBarView!!.addItem("Tasks", R.drawable.ic_baseline_assignment_24) {
            TasksDialog(this).show(
                parentFragmentManager, ""
            )
        }
        bottomBarView!!.addItem("Search", R.drawable.ic_round_search_24) {
            val searchFragment = SearchDialog(this, currentDirectory!!)
            searchFragment.show(parentFragmentManager, "")
            setSelectAll(false)
        }
        bottomBarView!!.addItem(
            "Create",
            R.drawable.ic_baseline_add_24
        ) { showAddNewFileDialog() }
        bottomBarView!!.addItem(
            "Sort",
            R.drawable.ic_baseline_sort_24
        ) { view: View -> showSortOptionsMenu(view) }
        bottomBarView!!.addItem(
            "Select All",
            R.drawable.ic_baseline_select_all_24
        ) { setSelectAll(true) }
        bottomBarView!!.addItem(
            "refresh",
            R.drawable.ic_baseline_restart_alt_24
        ) { refresh() }
    }

    private fun showSortOptionsMenu(view: View) {
        val popupMenu = PopupMenu(requireActivity(), view)
        popupMenu.menu.add("Sort by:").isEnabled = false
        popupMenu.menu.add("Name (A-Z)").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.sortingMethod == PrefsUtils.SORT_NAME_A2Z
        popupMenu.menu.add("Name (Z-A)").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.sortingMethod == PrefsUtils.SORT_NAME_Z2A
        popupMenu.menu.add("Size (Bigger)").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.sortingMethod == PrefsUtils.SORT_SIZE_BIGGER
        popupMenu.menu.add("Size (Smaller)").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.sortingMethod == PrefsUtils.SORT_SIZE_SMALLER
        popupMenu.menu.add("Date (Newer)").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.sortingMethod == PrefsUtils.SORT_DATE_NEWER
        popupMenu.menu.add("Date (Older)").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.sortingMethod == PrefsUtils.SORT_DATE_OLDER
        popupMenu.menu.add("Other options:").isEnabled = false
        popupMenu.menu.add("Folders first").setCheckable(true).isChecked =
            PrefsUtils.FileExplorerTab.listFoldersFirst()
        popupMenu.setOnMenuItemClickListener { menuItem: MenuItem ->
            menuItem.isChecked = !menuItem.isChecked
            when (menuItem.title.toString()) {
                "Name (A-Z)" -> {
                    PrefsUtils.FileExplorerTab.sortingMethod = PrefsUtils.SORT_NAME_A2Z
                }
                "Name (Z-A)" -> {
                    PrefsUtils.FileExplorerTab.sortingMethod = PrefsUtils.SORT_NAME_Z2A
                }
                "Size (Bigger)" -> {
                    PrefsUtils.FileExplorerTab.sortingMethod = PrefsUtils.SORT_SIZE_BIGGER
                }
                "Size (Smaller)" -> {
                    PrefsUtils.FileExplorerTab.sortingMethod = PrefsUtils.SORT_SIZE_SMALLER
                }
                "Date (Older)" -> {
                    PrefsUtils.FileExplorerTab.sortingMethod = PrefsUtils.SORT_DATE_OLDER
                }
                "Date (Newer)" -> {
                    PrefsUtils.FileExplorerTab.sortingMethod = PrefsUtils.SORT_DATE_NEWER
                }
                "Folders first" -> {
                    PrefsUtils.FileExplorerTab.setListFoldersFirst(menuItem.isChecked)
                }
            }
            refresh()
            true
        }
        popupMenu.show()
    }

    @SuppressLint("SetTextI18n")
    private fun showSetPathDialog(): Boolean {
        val customDialog = CustomDialog()
        val input = customDialog.createInput(requireActivity(), "File path")
        input.editText?.setSingleLine()
        val textView = MaterialTextView(requireContext())
        textView.setPadding(0, 8.toDp(), 0, 0)
        textView.alpha = 0.7f
        textView.text = "Quick Links:"
        val layout = ChipGroup(requireContext())
        val internal = Chip(requireContext())
        internal.layoutParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        internal.text = "Internal Data Files"
        internal.setOnClickListener {
            setCurrentDirectory(requireActivity().filesDir)
            customDialog.dismiss()
        }
        val external = Chip(requireContext())
        external.layoutParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        external.text = "External Data Files"
        external.setOnClickListener {
            setCurrentDirectory(requireActivity().getExternalFilesDir(null)!!)
            customDialog.dismiss()
        }
        layout.addView(internal)
        layout.addView(external)
        customDialog.setTitle("Jump to path")
            .addView(input)
            .addView(textView)
            .addView(layout)
            .setPositiveButton("Go", {
                val file = File(
                    input.editText!!.text.toString()
                )
                if (file.exists()) {
                    if (file.canRead()) {
                        if (file.isFile) {
                            FileOpener(requireActivity() as MainActivity).openFile(file)
                        } else {
                            setCurrentDirectory(file)
                        }
                    } else {
                        showMsg(Log.UNABLE_TO + " read the provided file")
                    }
                } else {
                    showMsg("The destination path doesn't exist!")
                }
            }, true)
            .show(parentFragmentManager, "")
        return true
    }

    private fun showAddNewFileDialog() {
        val customDialog = CustomDialog()
        val input = customDialog.createInput(requireActivity(), "File name")

        input.editText?.setSingleLine()
        FileUtils.setFileValidator(input, currentDirectory!!)
        CustomDialog()
            .setTitle("Create new file")
            .addView(input)
            .setPositiveButton("File", {
                createFile(
                    input.editText!!.text.toString(), false
                )
            }, true)
            .setNegativeButton("Folder", {
                createFile(
                    input.editText!!.text.toString(), true
                )
            }, true)
            .setNeutralButton("Cancel", null, true)
            .show(parentFragmentManager, "")
    }

    private fun createFile(name: String, isFolder: Boolean) {
        val file = File(currentDirectory, name)
        if (isFolder) {
            if (!file.mkdir()) {
                showMsg(Log.UNABLE_TO + " create folder: " + file.absolutePath)
            } else {
                refresh()
                focusOn(file)
            }
        } else {
            try {
                if (!file.createNewFile()) {
                    showMsg(Log.UNABLE_TO + " " + FileUtils.CREATE_FILE + ": " + file.absolutePath)
                } else {
                    refresh()
                    focusOn(file)
                }
            } catch (e: IOException) {
                Log.e(TAG, e)
                showMsg(e.toString())
            }
        }
    }

    override fun closeTab() {
        // Close the tab (if not default tab)
        if (tag != null && !tag!!.startsWith("0_")) {
            super.closeTab()
        }
    }

    override fun onBackPressed(): Boolean {
        //  Unselect selected files (if any)
        if (selectedFiles.size > 0) {
            setSelectAll(false)
            return true
        }
        // Go back if possible
        val parent = currentDirectory?.parentFile
        if (parent != null && parent.exists() && parent.canRead()) {
            setCurrentDirectory(currentDirectory?.parentFile!!)
            // restore RecyclerView state
            restoreRecyclerViewState()
            // Clear any selected files from the DataHolder (it also gets cleared
            // when a FileItem is clicked)
            (dataHolder as FileExplorerTabDataHolder).selectedFiles.clear()
            return true
        }
        // Close the tab (if not default tab)
        if (tag != null && !tag!!.startsWith("0_")) {
            closeTab()
            return true
        }
        return false
    }

    override fun createNewDataHolder(): BaseDataHolder {
        val dataHolder = FileExplorerTabDataHolder(tag!!)
        dataHolder.activeDirectory =
            if (currentDirectory == null) defaultHomeDirectory else currentDirectory
        return dataHolder
    }

    override fun onStop() {
        super.onStop()
        requireRefresh = true
    }

    override fun onPause() {
        super.onPause()
        requireRefresh = true
    }

    override fun onResume() {
        super.onResume()
        if (requireRefresh) {
            requireRefresh = false
            refresh()
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setSelectAll(select: Boolean) {
        if (!select) (dataHolder as FileExplorerTabDataHolder).selectedFiles.clear()
        for (item in files) {
            item.isSelected = select
            if (select) {
                (dataHolder as FileExplorerTabDataHolder).selectedFiles.add(item.file)
            }
        }
        // Don't call refresh(), because it will recreate the tab and reset the selection
        fileList.adapter?.notifyDataSetChanged()
    }

    val selectedFiles: ArrayList<FileItem>
        get() {
            val list = ArrayList<FileItem>()
            for (item in files) {
                if (item.isSelected) list.add(item)
            }
            return list
        }

    /**
     * Show/Hide placeholder
     */
    fun showPlaceholder(isShow: Boolean) {
        placeHolder.visibility = if (isShow) View.VISIBLE else View.GONE
    }

    /**
     * Used to update the title of attached tabView
     */
    private fun updateTabTitle() {
        getTabView()!!.setName(name)
    }

    /**
     * This method is called once from #onViewCreated(View, Bundle)
     */
    private fun initFileList() {
        fileList.adapter = FileListAdapter(this)
        fileList.setHasFixedSize(true)
        initPathHistory()
    }

    private fun initPathHistory() {
        pathHistoryRv.layoutManager =
            LinearLayoutManager(requireActivity(), LinearLayoutManager.HORIZONTAL, false)
        pathHistoryRv.adapter = PathHistoryAdapter(this)
    }

    /**
     * RecyclerView state should be saved when the fragment is destroyed and recreated.
     * #getDataHolder() isn't used here because we don't want to create a new DataHolder if the fragment is about
     * to close (note that the DataHolder gets removed just right before the fragment is closed)
     */
    override fun onDestroy() {
        super.onDestroy()
        val fileExplorerTabDataHolder = mainViewModel!!.getDataHolder(
            tag!!
        ) as FileExplorerTabDataHolder?
        fileExplorerTabDataHolder?.recyclerViewStates?.put(
            currentDirectory!!, fileList.layoutManager!!
                .onSaveInstanceState()!!
        )
    }

    /**
     * This method handles the following (in order):
     * - Updating currentDirectory and previousDirectory fields
     * - Updating recyclerViewStates in DataHolder
     * - Sorting files based on the preferences
     * - Updating tabView title
     * - Update pathHistory list
     * - Refreshing adapters (fileList & pathHistory)
     * - Updating activeDirectory in DataHolder
     *
     * @param file the directory to open
     */
    fun setCurrentDirectory(file: File) {
        if (currentDirectory != null) previousDirectory = currentDirectory
        currentDirectory = file
        // Save only when previousDirectory is set (so that it can restore the state before onDestroy())
        if (previousDirectory != null) (dataHolder as FileExplorerTabDataHolder).recyclerViewStates[previousDirectory!!] =
            fileList.layoutManager!!
                .onSaveInstanceState()!!
        prepareFiles()
        updateTabTitle()
        updatePathHistoryList()
        refreshFileList()
        (dataHolder as FileExplorerTabDataHolder).activeDirectory = currentDirectory
    }

    /**
     * This method automatically removes the restored state from DataHolder recyclerViewStates
     * This method is called when:
     * - Create the fragment
     * - #onBackPressed()
     * - when select a directory from pathHistory RecyclerView
     */
    fun restoreRecyclerViewState() {
        val savedState =
            (dataHolder as FileExplorerTabDataHolder).recyclerViewStates[currentDirectory]
        if (savedState != null) {
            fileList.layoutManager?.onRestoreInstanceState(savedState)
            (dataHolder as FileExplorerTabDataHolder).recyclerViewStates.remove(currentDirectory)
        }
    }

    /**
     * Refreshes both fileList and pathHistory recyclerview (used by #setCurrentDirectory(File) ONLY)
     */
    @SuppressLint("NotifyDataSetChanged")
    private fun refreshFileList() {
        fileList.adapter?.notifyDataSetChanged()
        pathHistoryRv.adapter?.notifyDataSetChanged()
        pathHistoryRv.scrollToPosition(pathHistoryRv.adapter!!.itemCount - 1)
        fileList.scrollToPosition(0)
        if (toolbar != null) toolbar!!.subtitle =
            currentDirectory?.getFormattedFileCount()
    }

    /**
     * Used to refresh the tab
     */
    fun refresh() {
        setCurrentDirectory(currentDirectory!!)
        restoreRecyclerViewState()
    }

    private val defaultHomeDirectory: File
        get() = Environment.getExternalStorageDirectory()

    private fun prepareFiles() {
        // Make sure current file is ready
        if (currentDirectory == null) {
            loadData()
            return
        }
        // Clear previous list
        files.clear()
        // Load all files in the current File
        val files = currentDirectory!!.listFiles()
        if (files != null) {
            for (comparator in FileUtils.comparators) {
                Arrays.sort(files, comparator)
            }
            for (file in files) {
                val fileItem = FileItem(file)
                if ((dataHolder as FileExplorerTabDataHolder).selectedFiles.contains(fileItem.file)) {
                    fileItem.isSelected = true
                }
                this.files.add(fileItem)
            }
        }
    }

    fun focusOn(file: File) {
        for (i in files.indices) {
            if (file == files[i].file) {
                fileList.scrollToPosition(i)
                return
            }
        }
    }

    /**
     * @return the name associated with this tab (currently used for tabView)
     */
    val name: String
        get() = currentDirectory?.getShortLabel(MAX_NAME_LENGTH)!!

    fun showFileOptions(fileItem: FileItem?) {
        if (!this::fileOptionsHandler.isInitialized) {
            fileOptionsHandler = FileOptionsHandler(this)
        }
        fileOptionsHandler.showOptions(fileItem!!)
    }

    fun openFile(fileItem: FileItem) {
        FileOpener(requireActivity() as MainActivity).openFile(fileItem.file)
    }

    fun showDialog(title: String?, msg: String?) {
        MaterialAlertDialogBuilder(requireContext())
            .setTitle(title!!)
            .setMessage(msg!!)
            .setPositiveButton("Ok", null)
            .show()
    }

    private fun updatePathHistoryList() {
        val list = ArrayList<File>()
        var file = currentDirectory
        while (file != null && file.canRead()) {
            list.add(file)
            file = file.parentFile
        }
        list.reverse()
        if (list.size > 0) pathHistory = list
    }

    companion object {
        const val MAX_NAME_LENGTH = 26
        private const val TAG = "FileExplorerTabFragment"
    }
}